File: lla2ecef_comparison/lla_to_ecef.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <stdbool.h>

#define GS_ACDS_deg2rad 0.01745329252

void lla_to_ecef(float lat, float lon, float alt, float *x, float *y, float *z);

void lla_to_ecef(float lat, float lon, float alt, float *x, float *y, float *z){
    
    float R_oblat;
    lat = lat * GS_ACDS_deg2rad;
    lon = lon * GS_ACDS_deg2rad;
    
    // WGS84 ellipsoid constants:
    float a = 6378137.0;
    float e = 8.1819190842622e-2;

    // intermediate calculation
    // (prime vertical radius of curvature)
    R_oblat = a / sqrt(1.0 - powf(e, 2.0) * powf(sin(lat), 2.0));
    
    *x = (R_oblat + alt) * cos(lat)*cos(lon);
    *y = (R_oblat + alt) * cos(lat)*sin(lon);
    *z = ((1.0 - powf(e, 2.0)) * R_oblat + alt) * sin(lat);
}

int main(int argc, char *argv[]) {
    if (argc != 4) {
        fprintf(stderr, "Usage: %s <latitude> <longitude> <altitude>\n", argv[0]);
        return 1;
    }

    float lat = atof(argv[1]);
    float lon = atof(argv[2]);
    float alt = atof(argv[3]);
    float x, y, z;
    
    lla_to_ecef(lat, lon, alt, &x, &y, &z);
    
    printf("{\"x\": %.6f, \"y\": %.6f, \"z\": %.6f}\n", x, y, z);
    return 0;
}

File: lla2ecef_comparison/lla_to_ecef.py
import sys
import json
from pyproj import Proj, transform

def lla_to_ecef(lat, lon, alt):
    # Define the projection for WGS84
    wgs84 = Proj(proj='latlong', datum='WGS84')
    
    # Define the projection for ECEF
    ecef = Proj(proj='geocent', datum='WGS84')
    
    # Transform coordinates from LLA to ECEF
    x, y, z = transform(wgs84, ecef, lon, lat, alt)
    
    return x, y, z

if len(sys.argv) != 4:
    print("Usage: python3 lla_to_ecef.py <latitude> <longitude> <altitude>")
    sys.exit(1)

lat = float(sys.argv[1])
lon = float(sys.argv[2])
alt = float(sys.argv[3])

x, y, z = lla_to_ecef(lat, lon, alt)

print(json.dumps({"x": x, "y": y, "z": z}, indent=2))